File: CsharpProjects/myLeetCode/Greedy/Sequential_problem/738. Monotone Increasing Digits.cs
// 738. Monotone Increasing Digits
// https://leetcode.com/problems/monotone-increasing-digits/

#include <iostream>
#include <string>

int monotoneIncreasingDigits(int n) {
    std::string s = std::to_string(n);

    int i = s.size() - 1;
    while (i > 0 && s[i] < s[i-1]) {
        s[i-1]--;
        i--;
    }

    for (int j = i+1; j < s.size(); j++) {
        s[j] = '9';
    }

    return std::stoi(s);
}

int main() {
    std::cout << "The largest number with monotone increasing digits is: " << monotoneIncreasingDigits(10) << std::endl;

    std::cout << "The largest number with monotone increasing digits is: " << monotoneIncreasingDigits(1234) << std::endl;

    std::cout << "The largest number with monotone increasing digits is: " << monotoneIncreasingDigits(332) << std::endl;

    return 0;
}

File: C++/代码随想录/Greedy/Sequential_problem/738. Monotone Increasing Digits.py
def monotoneIncreasingDigits(N):
    digits = list(str(N))
    n = len(digits)
    mark = n

    for i in range(n - 1, 0, -1):
        if digits[i] < digits[i - 1]:
            mark = i
            digits[i - 1] = str(int(digits[i - 1]) - 1)

    for i in range(mark, n):
        digits[i] = '9'

    return int(''.join(digits))

# 测试用例
print(monotoneIncreasingDigits(10))  # 应该输出 9
print(monotoneIncreasingDigits(1234))  # 应该输出 1234
print(monotoneIncreasingDigits(332))  # 应该输出 299
